### ISSUE TITLE
#[5466]: 期权波动率曲面深度学习 | Volatility Surface Deep Learning

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4350]
**标签**: quantitative-strategy, options, deep-learning, volatility-surface

### 摘要
利用深度学习模型（SABR-HJM、局部波动率模型）拟合期权波动率曲面，预测隐含波动率。

### 算法描述

**核心原理**

波动率曲面是期权定价的核心输入，由隐含波动率随执行价格和到期时间的三维分布构成。传统SABR模型虽能拟合波动率微笑，但参数校准耗时且不擅长外推。深度学习方法通过神经网络逼近从期权观测数据到隐含波动率的映射关系，实现实时预测。

**数学基础**

1. **SABR模型**：描述波动率动态
   $$dF(t) = \sigma(t) F(t)^\beta dW_1$$
   $$d\sigma(t) = \nu \sigma(t)^\gamma dW_2$$
   其中 $\beta \in [0,1]$ 控制微笑曲率，$\gamma$ 为波动率波动率，$\nu$ 为随机波动率强度。

2. **局部波动率模型**（Dupire方程）：
   $$\sigma_{loc}^2(K,T) = \frac{\partial C / \partial T}{K^2 \partial^2 C / \partial K^2}$$
   通过期权价格对执行价格和到期的偏导数反推局部波动率。

3. **神经网络结构**：
   - 输入层：$(K, T, r, F, TTE)$ 标准化的执行价格、剩余到期时间、无风险利率、前向价格、TTE
   - 隐藏层：4层全连接，每层256个神经元，ReLU激活，Dropout=0.2
   - 输出层：隐含波动率预测值 $\hat{sigma}$

**计算流程**

1. **数据采集**：收集期权市场报价（calls/puts），提取 $C(K,T)$ 价格矩阵
2. **隐含波动率反推**：使用Black-Scholes公式数值求解 $sigma_{imp}$
3. **特征工程**：构建 $(log(K/F), T, r, F, sqrt(T))$ 五元组
4. **模型训练**：最小化均方误差损失 $L = \sum (sigma_{pred} - sigma_{imp})^2$
5. **预测输出**：给定 $(K, T)$ 输出预测隐含波动率

### Python代码示例

```python
import numpy as np
import pandas as pd
from scipy.stats import norm
from scipy.optimize import brentq
from sklearn.model_selection import train_test_split
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader, TensorDataset


class BlackScholes:
    """Black-Scholes期权定价与隐含波动率计算"""
    
    @staticmethod
    def call_price(F, K, T, r, sigma):
        if T <= 0 or sigma <= 0:
            return max(F - K, 0)
        d1 = (np.log(F / K) + 0.5 * sigma**2 * T) / (sigma * np.sqrt(T))
        d2 = d1 - sigma * np.sqrt(T)
        return F * norm.cdf(d1) - K * np.exp(-r * T) * norm.cdf(d2)
    
    @staticmethod
    def put_price(F, K, T, r, sigma):
        return BlackScholes().call_price(F, K, T, r, sigma) - F + K * np.exp(-r * T)
    
    @staticmethod
    def implied_volatility(F, K, T, r, price, option_type='call', tol=1e-6):
        """利用Brentq方法反推隐含波动率"""
        def objective(sigma):
            if option_type == 'call':
                return BlackScholes().call_price(F, K, T, r, sigma) - price
            return BlackScholes().put_price(F, K, T, r, sigma) - price
        
        try:
            return brentq(objective, 1e-6, 10.0, xtol=tol)
        except:
            return np.nan


class VolatilitySurfaceDataset:
    """波动率曲面数据集构建"""
    
    def __init__(self, n_samples=10000):
        self.n_samples = n_samples
        self.data = self._generate_synthetic_data()
    
    def _generate_synthetic_data(self):
        """生成合成期权数据模拟真实市场"""
        np.random.seed(42)
        data = []
        
        for _ in range(self.n_samples):
            F = np.random.uniform(2500, 5000)
            r = np.random.uniform(0.01, 0.05)
            sigma_base = np.random.uniform(0.1, 0.4)
            
            for T in [0.1, 0.25, 0.5, 1.0]:
                for moneyness in np.linspace(0.8, 1.2, 9):
                    K = F * moneyness
                    smile = 0.02 * (1 - np.exp(-(moneyness - 1)**2 / 0.1))
                    sigma = sigma_base * (1 + smile)
                    T_noise = max(T + np.random.normal(0, 0.01), 0.01)
                    price = BlackScholes().call_price(F, K, T_noise, r, sigma)
                    price *= (1 + np.random.normal(0, 0.005))
                    iv = BlackScholes().implied_volatility(F, K, T_noise, r, price)
                    
                    if not np.isnan(iv) and 0.01 < iv < 2.0:
                        data.append({
                            'F': F, 'K': K, 'T': T, 'r': r,
                            'log_money': np.log(K / F),
                            'TTE': np.sqrt(T),
                            'sigma_imp': iv
                        })
        
        return pd.DataFrame(data)
    
    def get_tensors(self, test_size=0.2):
        X = self.data[['log_money', 'T', 'r', 'F', 'TTE']].values
        y = self.data['sigma_imp'].values.reshape(-1, 1)
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=test_size, random_state=42)
        return (X_train, y_train), (X_test, y_test)


class VolatilitySurfaceNet(nn.Module):
    """深度学习波动率曲面模型"""
    
    def __init__(self, input_dim=5, hidden_dim=256, dropout=0.2):
        super().__init__()
        self.network = nn.Sequential(
            nn.Linear(input_dim, hidden_dim),
            nn.ReLU(),
            nn.Dropout(dropout),
            nn.Linear(hidden_dim, hidden_dim),
            nn.ReLU(),
            nn.Dropout(dropout),
            nn.Linear(hidden_dim, hidden_dim),
            nn.ReLU(),
            nn.Dropout(dropout),
            nn.Linear(hidden_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, 1)
        )
    
    def forward(self, x):
        return self.network(x)


def train_model(model, train_loader, optimizer, criterion, epochs=100):
    model.train()
    for epoch in range(epochs):
        total_loss = 0
        for X_batch, y_batch in train_loader:
            optimizer.zero_grad()
            pred = model(X_batch)
            loss = criterion(pred, y_batch)
            loss.backward()
            optimizer.step()
            total_loss += loss.item()
        if (epoch + 1) % 20 == 0:
            print(f"Epoch {epoch+1}, Loss: {total_loss / len(train_loader):.6f}")


def evaluate_model(model, X_test, y_test):
    model.eval()
    with torch.no_grad():
        y_pred = model(torch.FloatTensor(X_test)).numpy()
    mse = np.mean((y_pred - y_test)**2)
    rmse = np.sqrt(mse)
    print(f"Test RMSE: {rmse:.6f}")
    return y_pred


def main():
    print("生成波动率曲面数据集...")
    dataset = VolatilitySurfaceDataset(n_samples=5000)
    (X_train, y_train), (X_test, y_test) = dataset.get_tensors()
    print(f"训练集: {len(X_train)}, 测试集: {len(X_test)}")
    
    train_dataset = TensorDataset(torch.FloatTensor(X_train), torch.FloatTensor(y_train))
    train_loader = DataLoader(train_dataset, batch_size=256, shuffle=True)
    
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    model = VolatilitySurfaceNet().to(device)
    optimizer = optim.Adam(model.parameters(), lr=0.001)
    criterion = nn.MSELoss()
    
    print("训练模型...")
    train_model(model, train_loader, optimizer, criterion, epochs=150)
    
    print("评估模型...")
    y_pred = evaluate_model(model, X_test, y_test)
    
    print("\\n预测示例:")
    for strike in [0.9, 1.0, 1.1]:
        F, K, T, r = 3500, 3500 * strike, 0.5, 0.03
        log_money = np.log(K / F)
        features = np.array([[log_money, T, r, F, np.sqrt(T)]])
        with torch.no_grad():
            pred_sigma = model(torch.FloatTensor(features)).item()
        print(f"  K/F={strike:.1f}, T={T}y: 预测IV={pred_sigma:.4f}")


if __name__ == "__main__":
    main()
```

### 数据来源
- Reddit: r/quant, r/algorithmictrading, r/Finance
- 小红书: 量化投资、理财技巧
- 知乎: 量化交易、机器学习交易
- 其他: Investopedia, Quantopian, arXiv
